### PR TITLE
openshift_checks: enable variable conversion

### PR DESCRIPTION
**Pull request by sosiouxme**
get_var can now do type conversion and raise an error if it doesn't work.

Later this could be integrated with defined check parameters to give the user more info about what was expected. Also check parameters could have indications for whether they're deprecated and defined/used.

Finally somewhere in here we could deal with the lazy evaluation of jinja2 in variables.

### CODE REVIEW

File: roles/openshift_health_checker/openshift_checks/__init__.py [hunk 1]
```diff
@@ -97,12 +97,39 @@ def get_var(self, *keys, **kwargs):
...
+            keys = keys[0].split(".")
+
         try:
             value = reduce(operator.getitem, keys, self.task_vars)
         except (KeyError, TypeError):
             if "default" in kwargs:
                 return kwargs["default"]
             raise OpenShiftCheckException("'{}' is undefined".format(".".join(map(str, keys))))
+
+        convert = kwargs.get("type")
```
Comment (rhcarvalho) [inline roles/openshift_health_checker/openshift_checks/__init__.py]: Suggestion:

```python
convert = kwargs.get("type", lambda x: x)
try:
    return convert(value)
except ValueError:
    raise OpenShiftCheckException(...)
```

Wherever we use `.get_var(...)`, we'd then always pass a function to the `type` optional argument, and never a string. Passing a string would be more cumbersome anyway.

File: roles/openshift_health_checker/test/openshift_check_test.py [hunk 2]
```diff
@@ -81,6 +81,7 @@ class TestCheck(OpenShiftCheck):
 @pytest.mark.parametrize("keys,expected", [
     (("foo",), 42),
     (("bar", "baz"), "openshift"),
+    (("bar.baz",), "openshift"),
```
Comment (rhcarvalho) [inline roles/openshift_health_checker/test/openshift_check_test.py]: IIRC we had this before and then removed the "extra feature" in favor of a single way of doing it.
Are you intending to use both `get_var('openshift.foo.bar')` and `get_var('openshift', 'foo', 'bar')`?

File: roles/openshift_health_checker/test/openshift_check_test.py [hunk 3]
```diff
@@ -94,3 +95,22 @@ def test_get_var_error(task_vars, missing_keys):
...
+    (("bar", "baz"), bool, False),
+])
+def test_get_var_convert(task_vars, keys, convert, expected):
+    assert dummy_check(task_vars).get_var(*keys, type=convert) == expected
+
+
+@pytest.mark.parametrize("keys, convert", [
+    (("foo",), "bogus"),
+    (("bar", "baz"), "int"),
+    (("bar.baz"), "float"),
```
Comment (rhcarvalho) [inline roles/openshift_health_checker/test/openshift_check_test.py]: Funny that this test uses all strings, and the other uses all functions as the `convert` argument. I think dropping strings altogether makes the design and usage simpler.

File: roles/openshift_health_checker/openshift_checks/__init__.py (hunk 1 above)
Comment (sosiouxme) [inline roles/openshift_health_checker/openshift_checks/__init__.py]: That seems like a good idea. I was thinking we might want logic more complicated to e.g. extract version strings or convert byte suffixes or something, but I guess passing a function would cover that too.

Comment (sosiouxme) [inline roles/openshift_health_checker/openshift_checks/__init__.py]: although, note the logic under bool... part of the reason for doing this is to not have to repeat that everywhere.

I suppose we could still special-case known conversion functions like bool though.

Comment (rhcarvalho) [inline roles/openshift_health_checker/openshift_checks/__init__.py]: > note the logic under bool...

Hmm, why would we add that YAML-like conversion here?

For `bool`, any non-empty string converts to `True`.

Now this gets me thinking that "covert" is normally from A to B, can we say that A is always of type `str`?

File: roles/openshift_health_checker/test/openshift_check_test.py (hunk 2 above)
Comment (sosiouxme) [inline roles/openshift_health_checker/test/openshift_check_test.py]: Why not? It's two extra lines and it matches what we do elsewhere in ansible code. Personally I would like the dotted version better but I don't want to go through the code and change them all.

I'm not as set on the "single obvious way to do it" ethos of python...

File: roles/openshift_health_checker/openshift_checks/__init__.py (hunk 1 above)
Comment (sosiouxme) [inline roles/openshift_health_checker/openshift_checks/__init__.py]: bool("false") is True. Ansible inventory variables typically come in as strings, except where we've set them with facts already. This has already caused multiple bugs. See the change in logging/logging.py - someone who set `openshift_hosted_deploy_logging=False` in their inventory would see the checks expecting logging to be deployed. It's a really easy mistake to make.

**Comment by sosiouxme** [comment]
@rhcarvalho updated

File: roles/openshift_health_checker/openshift_checks/__init__.py [hunk 4]
```diff
@@ -107,30 +107,27 @@ def get_var(self, *keys, **kwargs):
                 return kwargs["default"]
             raise OpenShiftCheckException("'{}' is undefined".format(".".join(map(str, keys))))
 
-        convert = kwargs.get("type")
+        convert = kwargs.get("convert", lambda x: x)
```
Comment (rhcarvalho) [inline roles/openshift_health_checker/openshift_checks/__init__.py]: Perhaps the default is really to have strings... in which case `str` as default would be a good deal.

File: roles/openshift_health_checker/openshift_checks/__init__.py [hunk 5]
```diff
@@ -107,30 +107,27 @@ def get_var(self, *keys, **kwargs):
...
-            if convert is None:
-                pass
-            elif convert in ["str", str]:
-                return str(value)
-            elif convert in ["int", int]:
-                return int(value)
-            elif convert in ["float", float]:
-                return float(value)
-            elif convert in ["bool", bool]:
+            if convert is bool:
```
Comment (rhcarvalho) [inline roles/openshift_health_checker/openshift_checks/__init__.py]: Suggestion: instead of changing what `bool` (the function) means behind the scenes, we could have an explicit conversion function that handles all the corner cases we want, that would be similar to how API like Django Models and other Python libraries work.

We'd essentially declare `type=SpecialBool` (surely name `SpecialBool` anything different than that...) for the variables where we want the special handling. That also leaves room for us to use `bool` in the future without breaking any then existing code.

File: roles/openshift_health_checker/openshift_checks/__init__.py [hunk 6]
```diff
@@ -107,30 +107,27 @@ def get_var(self, *keys, **kwargs):
...
-                )
+                return convert(value)
+
         except ValueError as error:
             raise OpenShiftCheckException(
-                'Cannot convert variable "{var}" to expected type "{type}":\n'
-                '{error}'.format(var=".".join(keys), type=convert, error=str(error))
+                'Cannot convert inventory variable to expected type:\n'
+                '  "{var}={value}"\n'
+                '{error}'.format(var=".".join(keys), value=value, error=str(error))
```
Comment (rhcarvalho) [inline roles/openshift_health_checker/openshift_checks/__init__.py]: I think the `str` around `error` is unnecessary?

File: roles/openshift_health_checker/openshift_checks/__init__.py [hunk 7]
```diff
@@ -107,30 +107,27 @@ def get_var(self, *keys, **kwargs):
...
             )
 
-        return value
+        except Exception as error:
+            raise OpenShiftCheckException(
+                'There is a bug in this check. While trying to convert variable \n'
+                '  "{var}={value}"\n'
+                'the given converter cannot be used or failed unexpectedly:\n'
+                '{error}'.format(var=".".join(keys), value=value, error=error)
+            )
```
Comment (rhcarvalho) [inline roles/openshift_health_checker/openshift_checks/__init__.py]: Hmm, this means that if the function returns something that is not a `ValueError` we want to report a slightly different message?!

File: roles/openshift_health_checker/openshift_checks/logging/logging.py [hunk 8]
```diff
@@ -15,7 +15,7 @@ class LoggingCheck(OpenShiftCheck):
     logging_namespace = "logging"
 
     def is_active(self):
-        logging_deployed = self.get_var("openshift_hosted_logging_deploy", type=bool, default=False)
+        logging_deployed = self.get_var("openshift_hosted_logging_deploy", convert=bool, default=False)
```
Comment (rhcarvalho) [inline roles/openshift_health_checker/openshift_checks/logging/logging.py]: `convert=bool` reads well, though `convert=str` is a bit strange since we perhaps expect the input to be a string already?! I liked `type`, though it is the name of a builtin function and may make pylint complain ;-)

File: roles/openshift_health_checker/openshift_checks/__init__.py (hunk 7 above)
Comment (sosiouxme) [inline roles/openshift_health_checker/openshift_checks/__init__.py]: `ValueError` ought to mean the user passed in a wrong value.

Any other kind of error indicates the conversion function writer did not handle the value correctly or at least gracefully (`KeyError` and similar). The user might have done something wrong too, but if it's not obvious to them then they can blame our code.

Are there other error categories that should be included / called out?

File: roles/openshift_health_checker/openshift_checks/logging/logging.py (hunk 8 above)
Comment (sosiouxme) [inline roles/openshift_health_checker/openshift_checks/logging/logging.py]: "type" is nice and intuitive for the most of what we're doing now with values coming out of `task_args`. I don't think pylint complains since it's never used as a name. But I think we may come up with more interesting standardized conversion and validation functions that don't just translate to a type. For instance, byte size converters, or validation that `deployment_type` is one of the known values.

So it made sense to me to give it a more general name that doesn't imply a simple typecasting. I'm not set on `convert` though, suggestions welcome... I could easily see `make`, `transform`, `expect`, `ensure` ... or accepting multiple names (`type`, `convert`, `validate`), to support not having to remember which it is :)

File: roles/openshift_health_checker/openshift_checks/__init__.py (hunk 7 above)
Comment (sosiouxme) [inline roles/openshift_health_checker/openshift_checks/__init__.py]: Now that I think about it, OpenShiftCheckException (from any custom converters we write) should probably just be passed through.

File: roles/openshift_health_checker/openshift_checks/__init__.py (hunk 4 above)
Comment (sosiouxme) [inline roles/openshift_health_checker/openshift_checks/__init__.py]: There are cases where the user can pass in JSON - disk availability for instance... but it might be best anyway to be explicit about expected values for those.

File: roles/openshift_health_checker/openshift_checks/__init__.py [hunk 9]
```diff
@@ -107,30 +107,27 @@ def get_var(self, *keys, **kwargs):
...
-                pass
-            elif convert in ["str", str]:
-                return str(value)
-            elif convert in ["int", int]:
-                return int(value)
-            elif convert in ["float", float]:
-                return float(value)
-            elif convert in ["bool", bool]:
+            if convert is bool:
                 return str(value).lower() in ["true", "yes", "1"]
```
Comment (sosiouxme) [inline roles/openshift_health_checker/openshift_checks/__init__.py]: I forgot about `y|n` and `on|off`.

Ansible is sort of a mishmash of ini and yaml, since you have the inventory as well as code, group variable files, etc.

My intuition is less about following a standard and more about not surprising the users. Which is a little bit chicken and egg, since following a standard should lead to consistent user expectations, in theory.

In the end, we should probably not implement our own here, and just re-use whatever the jinja `bool` filter already does. Then it will at least be consistent with what the rest of Ansible does, and "not our problem."

File: roles/openshift_health_checker/openshift_checks/__init__.py (hunk 5 above)
Comment (sosiouxme) [inline roles/openshift_health_checker/openshift_checks/__init__.py]: I really doubt there is an actual use case for using the built-in `bool` as the converter. In this context it is essentially `non_empty` so calling it `bool` would be highly confusing and lead to bugs.

File: roles/openshift_health_checker/openshift_checks/__init__.py (hunk 7 above)
Comment (rhcarvalho) [inline roles/openshift_health_checker/openshift_checks/__init__.py]: > Now that I think about it, OpenShiftCheckException (from any custom converters we write) should probably just be passed through.

I hope we will not get to that? I thought the conversion functions would be more like pure-Python simple functions, just to operate on basic data types.

The `ValueError` + others separation is good, we'll just need to keep it in mind and let our custom functions raise ValueError on bad input and not something else accidental like you mention `KeyError` and similar. Again, I hope perhaps there will be nothing custom apart from the boolean conversion?

File: roles/openshift_health_checker/openshift_checks/logging/logging.py (hunk 8 above)
Comment (rhcarvalho) [inline roles/openshift_health_checker/openshift_checks/logging/logging.py]: > or accepting multiple names (type, convert, validate), to support not having to remember which it is :)

Hahahahha :rofl:

File: roles/openshift_health_checker/openshift_checks/__init__.py (hunk 9 above)
Comment (rhcarvalho) [inline roles/openshift_health_checker/openshift_checks/__init__.py]: If it would be possible to delegate/reuse any part of Ansible, jinja, ruamel.yaml, etc, it would be advantageous for us :+1:

File: roles/openshift_health_checker/openshift_checks/__init__.py (hunk 4 above)
Comment (rhcarvalho) [inline roles/openshift_health_checker/openshift_checks/__init__.py]: Oh my, even JSON... I forgot about that ;)

The best would be to exterminate the INI format and command line flags in the next Ansible version, and made convert it all to Haskell -- hhaha, or maybe not :P

File: roles/openshift_health_checker/openshift_checks/__init__.py (hunk 9 above)
Comment (sosiouxme) [inline roles/openshift_health_checker/openshift_checks/__init__.py]: Ansible [implements its own bool filter](https://github.com/ansible/ansible/blob/5f73bdc3bf3821c957cddbf81febfc13b3de9af9/lib/ansible/plugins/filter/core.py#L117-L125). It's not exactly a standard but... we can be consistent.

File: roles/openshift_health_checker/openshift_checks/logging/logging.py (hunk 8 above)
Comment (sosiouxme) [inline roles/openshift_health_checker/openshift_checks/logging/logging.py]: Still open to other names... `normalize`... `coerce` ....

File: roles/openshift_health_checker/openshift_checks/__init__.py (hunk 7 above)
Comment (sosiouxme) [inline roles/openshift_health_checker/openshift_checks/__init__.py]: I could see other converters being useful; code we already have could use a standard byte size converter and `deployment_type` validator. I don't think it's too much scope creep / YAGNI to just reraise `OpenShiftCheckException`.

**Comment by sosiouxme** [comment]
aos-ci-test

**Comment by sosiouxme** [comment]
@rhcarvalho you ok with this? Can I get a LGTM?

File: roles/openshift_health_checker/openshift_checks/logging/logging.py [hunk 10]
```diff
@@ -18,7 +18,7 @@ class LoggingCheck(OpenShiftCheck):
     logging_namespace = "logging"
 
     def is_active(self):
-        logging_deployed = self.get_var("openshift_hosted_logging_deploy", default=False)
+        logging_deployed = self.get_var("openshift_hosted_logging_deploy", convert=bool, default=False)
```
Comment (rhcarvalho) [inline roles/openshift_health_checker/openshift_checks/logging/logging.py]: Is this fixing an existing bug?

Comment (sosiouxme) [inline roles/openshift_health_checker/openshift_checks/logging/logging.py]: I don't think I filed a bug. But yeah, it's a minor bug that logging is considered always on if defined...

**Comment by sosiouxme** [comment]
[merge]

**Comment by sosiouxme** [comment]
yum flakes https://github.com/openshift/origin/issues/10162 and https://github.com/openshift/origin/issues/8571

[merge]

**Comment by sosiouxme** [comment]
https://ci.openshift.redhat.com/jenkins/job/merge_pull_request_openshift_ansible/817/

yum flake https://github.com/openshift/origin/issues/8571
also https://github.com/openshift/origin/issues/11114

@rhcarvalho or @sdodson could you merge this manually please.
